test/permissions/remote.provider.create.js

```diff
@@ -0,0 +1,198 @@
+'use strict';
+
+const jstp = require('@metarhia/jstp');
+
+const { codes: errorCodes } = require('../../lib/errors');
+
+const EXPECTED_ERROR_CODE = errorCodes.INSUFFICIENT_PERMISSIONS;
+const EXPECTED_ERROR = new jstp.RemoteError(EXPECTED_ERROR_CODE);
+
+const isError = async (test, provider, category, record) => {
+  const error = await test.rejects(
+    () => provider.create(category, record),
+    EXPECTED_ERROR
+  );
+  test.strictSame(error.code, EXPECTED_ERROR_CODE);
+};
+
+module.exports = (test, { categories, users, roles, catalogs, subsystems }) => {
+  test.endAfterSubtests();
+
+  test.test(
+    'remoteProvider.create: not logged in',
+    async (test, { remoteProvider }) => {
+      const record = {
+        Caption: 'C4',
+        Category: categories.Document,
+      };
+
+      await isError(test, remoteProvider, 'Catalog', record);
+    }
+  );
+
+  test.test(
+    'remoteProvider.create: no such system user',
+    async (test, { remoteProvider, login }) => {
+      const record = {
+        Caption: 'C4',
+        Category: categories.Document,
+      };
+
+      await login(roles.S1.Admin);
+      await isError(test, remoteProvider, 'Catalog', record);
+    }
+  );
+
+  test.test(
+    'remoteProvider.create: blocked role',
+    async (test, { remoteProvider, login }) => {
+      const record = {
+        Caption: 'C4',
+        Category: categories.Document,
+      };
+
+      await login(users.S1.BlockedRoleUser);
+      await isError(test, remoteProvider, 'Catalog', record);
+    }
+  );
+
+  test.test(
+    'remoteProvider.create: no insert permission',
+    async (test, { remoteProvider, login }) => {
+      const record = {
+        Caption: 'C4',
+        Category: categories.Document,
+      };
+
+      await login(users.S1.UpdateOnlyUser);
+      await isError(test, remoteProvider, 'Catalog', record);
+    }
+  );
+
+  test.test(
+    'remoteProvider.create: no permission for category',
+    async (test, { remoteProvider, login }) => {
+      const record = {
+        Caption: 'C4',
+        Category: categories.Document,
+      };
+
+      await login(users.S2.InsertOnlyUser);
+      await isError(test, remoteProvider, 'Catalog', record);
+    }
+  );
+
+  test.test(
+    'remoteProvider.create: sufficient permission',
+    async (test, { remoteProvider, login }) => {
+      const record = {
+        Caption: 'C4',
+        Category: categories.Document,
+      };
+
+      await login(users.S1.InsertOnlyUser);
+      await test.resolves(async () => {
```

Comment:
```suggestion
      await test.resolves(remoteProvider.create('Catalog', record));
```

Reply:
@nechaido It wouldn't work since `test.resolves()` compares the result with the expected value, which is `undefined`, and `provider.create()` returns a value that is not `undefined`.

Reply:
This won't work, as `remoteProvider.create()` returns id and then `test.resolves()` does `strictSame` check, so I have to return nothing explicitly.